src/FnameResolver.sol

```diff
@@ -0,0 +1,157 @@
+
+    /**
+     * @dev Emit an event when the contract owner authorizes a new signer.
+     *
+     * @param signer Address of the authorized signer.
+     */
+    event AddSigner(address indexed signer);
+
+    /**
+     * @dev Emit an event when the contract owner removes an authorized signer.
+     *
+     * @param signer Address of the removed signer.
+     */
+    event RemoveSigner(address indexed signer);
+
+    /*//////////////////////////////////////////////////////////////
+                                CONSTANTS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Contract version. Follows Farcaster protocol version scheme.
+     */
+    string public constant VERSION = "2023.07.12";
+
+    /**
+     * @dev EIP-712 typehash of the UsernameProof struct.
+     */
+    bytes32 internal constant _USERNAME_PROOF_TYPEHASH =
+        keccak256("UsernameProof(string name,uint256 timestamp,address owner)");
+
+    /*//////////////////////////////////////////////////////////////
+                                STRUCTS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice An FIP-90 username proof.
+     *         See: https://github.com/farcasterxyz/protocol/discussions/90
+     */
+    struct UsernameProof {
+        string name;
+        uint256 timestamp;
+        address owner;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              PARAMETERS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev URL of the CCIP lookup gateway.
+     */
+    string public url;
+
+    /**
+     * @dev Mapping of signer address to authorized boolean.
+     */
+    mapping(address signer => bool isAuthorized) public signers;
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Set the lookup gateway URL, and initial signer.
+     *
+     * @param _url                     Lookup gateway URL. This value is set permanently.
+     * @param _signer                  Initial authorized signer address.
+     */
+    constructor(string memory _url, address _signer) EIP712("Farcaster name verification", "1") {
+        url = _url;
+        signers[_signer] = true;
+        emit AddSigner(_signer);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                             RESOLVER VIEWS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Resolve the provided ENS name. This function will always revert to indicate an offchain lookup.
+     */
+    function resolve(bytes calldata name, bytes calldata data) external view returns (bytes memory) {
+        bytes memory callData = abi.encodeCall(this.resolve, (name, data));
```

Comment:
I think we need to add an interface similar to this https://github.com/ensdomains/offchain-resolver/blob/8e25fd7a3420cfb898d27f40b044eb0b5f2c3a9d/packages/contracts/contracts/OffchainResolver.sol#L8-L10

where the return value is the UserNameProof struct (or the members of the struct if the struct is encoded differently

Reply:
Yes, you're right. I also forgot ERC-165 interface detection.